CIP-????/README.md

```diff
@@ -0,0 +1,709 @@
+```
+t ::=
+  ...
+
+  -- Tags a term with a tag
+  | (tag i t)
+  -- Inspects a tagged term for the tag, and passes the 
+  -- inner term to the case function corresponding to the tag
+  | (case t t...)
+
+  -- Constructs a product with the given fields
+  | (prod t...)
+  -- Accesses the i'th field of the given product
+  | (index i t)
+```
+
+This is cleaner in some ways, and was the first prototype we implemented, but it has the following problems:
+- It adds two more term constructors to the language
+- It is significantly slower in practice, because the combined operation of processing a sum-of-products is extremely common
+
+#### 2. Bind Case Variables in the Syntax for Case Branches
+
+The design presented here allows the case branches to be arbitrary terms, which must be evaluated and then applied to the fields.
+
+A more complex design would be to have the bindings for the variables be part of the case expression instead. That is, the current design does this:
+
+```
+(case (constr 1 a b) (\x -> t1) (\x y -> t2))
+```
+(the literal lambdas here could be arbitrary terms)
+
+whereas we could do this:
+
+```
+(case (constr 1 x y) (alt x t1) (alt x y t2))
+```
+
+The effect of this is that we know statically that we have a function, and we can jump right into evaluating the _body_ of the case branch instead of first having to evaluate it to a function and apply the arguments.
+
+This saves us a small but meaningful amount of overhead at the cost of making the implementation significantly more complex.
+
+We prototyped this version and it was noticeably faster (~10%).
+However, performance investigations showed that we can realize a significant amount of the performance gain through other means, leaving only about 3% unclaimed.[^see-appendix-2]
+We think that this is an acceptable cost for a simpler implementation.
+
+[^see-appendix-2]: See Appendix 2 for more details.
+
+#### 3. Convert Redeemers and Datums to Sums-of-Products Rather Than Builtin Data
+
+This proposal changes how the script context is passed to scripts, thus removing one use of the Plutus builtin type `Data`.
+However, we still use the builtin `Data` type for passing redeemers and datums, which continue to be `Data` objects in the transaction.[^why-redeemers-datums-data]
+
+[^why-redeemers-datums-data]: See "Why not also make datums and redeemers terms instead of `Data`?" for why we want to keep redeemers and datums as `Data` in transactions themselves.
+
+An alternative would be to encode the `Data` objects for redeemers and datums using sums-of-products instead of the builtin `Data` type.
+That would require us to define and specify a stable `dataToTerm :: Data -> Term` function that converts a `Data` object into a term using sums-of-products.
+This would be simple for some cases (e.g. `Constr`), but less simple for other cases (notably `Map` and `List`).
+
+This has some costs:
+1. We have an additional conversion function that we need to specify.
```

Comment:
If this function is defined to begin with, it can also be used for the script context. Which would highly simplify the implementation of that CIP in the ledger. All the ledger would need is to compose the existing translation interface with this new `dataToTerm` function to obtain a canonical Plutus `Term` representation of _any_ `Data`. This can be applied to datums, redeemers and the script context alike.

Reply:
Yes, although now we have to pay an additional translation cost every time. It's not clear to me that the consistency is worth the cost of not just doing it directly.

Reply:
I think it's more a development process strategy here. First, make it work. Then, make it fast. Given the current implementation, it's relatively easy to add a translation from `Data` to `Term` and unless I am missing something, I believe that conversation to be relatively trivial and _cheap_. So from a development perspective, it makes the change easy to implement and low-risk. 

Then, one can think in a second time how to make this more efficient by combining both steps and producing a `Term` directly when translating from ledger types.

Reply:
Yes that's fair. On reflection I think I'm more concerned about it being complex. And we'd have to specify it this way. I look forward to all the questions about "why is the script context translated in this weird two-step process?" :)

Reply:
Well, now you can link them to this comments thread 😬